### ISSUE TITLE
Checker for using 'is' with immutable objects

### ISSUE THREAD

**Issue by The-Compiler** [PyCQA/pylint#786]
One newbie mistake I see often is using `is` with immutable objects, like strings or integers. This can act like it does the right thing if you're lucky, and then break horribly with other values :wink:.

Types to (probably) complain about:

- `str`
- `unicode`
- `bytes`
- `int`
- `float`
- `long`
- `tuple`
- `complex`?
- `frozenset`?

(the ones marked with ? are immutable, but there are no literals for it - I think this means they'll never be interned, so at least `foo is bar` doesn't *sometimes* do the right thing)

Types where `foo is bar` is okay:

- Mutable types
- `None` should be okay, as `is None` is common
- `True` and `False` - while `if foo is True:` instead of `if foo:` is common, I think there are a few usecases for `is True` and `is False`.

**Comment by PCManticore** [comment]
Sounds good to me.

**Comment by xmo-odoo** [comment]
> * None should be okay, as is None is common
> * True and False - while if foo is True: instead of if foo: is common, I think there are a few usecases for is True and is False.

Basically, singleton types. An other case is sentinel `object` instances when a value/parameter can be an arbitrary object if user-provided e.g. (lifted straight from the standard library)
```python
__marker = object()
def pop(self, key, default=__marker):
    […]
    if default is self.__marker:
        # do thing when we know for sure no default value was provided
    else:
        # do whatever happens when a default value was provided
```

**Comment by PCManticore** [comment]
Just to leave this here, someone has to write it though. :D I won't probably have time in the following weeks.

**Comment by The-Compiler** [comment]
Yeah - I hope to have some more time for pylint in the somewhat-distant future (say, a few months) after the big things for qutebrowser are done.

Until then I'll just drop ideas here so I don't forget about them, and hope someone else will implement them before I get to it :wink:

**Comment by nedbat** [comment]
This talks about immutable types, but I'm not sure that's the heart of the problem.  I think it might be better to focus on literals.  I would want `x is [1, 2, 3]` to be flagged as well.

**Comment by PCManticore** [comment]
Pushed a basic implementation into the master branch, so it will be available for Pylint 2.0. I decided to warn about the literals, instead of making the distinction between mutable or immutable. The check does not do inference, so it means it will catch only the cases where we use a literal where it is defined, instead of catching cases such as ``x is y``, where x and y are names altogether. I restricted this mostly because this use case can be used for verifying a sentinel value, which pylint cannot understand right now, so it would have had false positives for cases where it shouldn't.
